## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [140]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [141]:
# View the column names in the data
df.columns



Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [142]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [143]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df['job'].nunique()
pd.options.display.max_rows = None
df['job'].value_counts()


job
Osteopath                                                      631
Toxicologist                                                   546
Architectural technologist                                     430
Librarian, academic                                            389
Teacher, special educational needs                             388
Environmental manager                                          364
Public relations account executive                             362
Call centre manager                                            355
Pension scheme manager                                         349
Audiological scientist                                         329
Emergency planning/management officer                          329
Race relations officer                                         325
Physiotherapist                                                305
Hydrographic surveyor                                          301
Prison officer                                            

In [144]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [145]:
# For the category with the most entries, which subcategory had the most entries?
consumables_df = df.loc[(df["category"]=="consumables"),:]
consumables_df["subcategory"].value_counts().head(1)


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [146]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [147]:
# Store the client ids of those top 5 clients in a list.
top_5 = df["client_id"].value_counts().head(5).index.to_list()
print(top_5)


[33615, 66037, 46820, 38378, 24741]


In [148]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = df["client_id"].value_counts().idxmax()
total_qty = df[df["client_id"]==top_client]["qty"].sum()
print(total_qty)


64313


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [149]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
subtotal = df["unit_price"] * df["qty"]
df["subtotal"]= subtotal
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [150]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calc_shipping(row):
    order_weight = row['unit_weight'] * row['qty']
    if order_weight > 50:
        return order_weight * 7
    return order_weight * 10

df['shipping_price'] = df.apply(calc_shipping, axis=1)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [151]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = df.apply(lambda row: round((row["subtotal"] + row["shipping_price"]) * 1.0925, 2), axis=1)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00         5512.50   
1      15.09         1.49   21            0     523.95          312.90   
2       7.86         1.68   39            6     527.28          458.64   
3      24.85         1.23   29            3    1056.18          356.70   
4     108.17        46.43   20            1    3902.00         6500.20   

   total_price  
0    131839.08  
1       914.26  
2      1077.12  
3      1543.57  
4     11364.40  

[5 rows x 21 columns]

In [152]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df["total_cost"] = df.apply(lambda row: round((row["qty"] * row["unit_cost"]) + row["shipping_price"],2),axis=1)
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price  total_price  \
0         7.50  105            1  115164.00         5512.50    131839.08   
1         1.49   21            0     523.95          312.90       914.26   
2         1.68   39            6     527.28          458.64      1077.12   
3         1.23   29            3    1056.18          356.70      1543.57   
4        46.43   20            1    3902.00         6500.20     11364.40   

   total_cost  
0    85597.05  
1      629.79  
2      765.18  
3     1077.35  
4     8663.60  

[5 rows x 22 columns]

In [153]:
# Create a column for the profit of each line using line cost and line price
df["profit"] = df["total_price"] - df["total_cost"]
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price  total_price  total_cost    profit  
0            1  115164.00         5512.50    131839.08    85597.05  46242.03  
1            0     523.95          312.90       914.26      629.79    284.47  
2            6     527.28          458.64      1077.12      765.18    311.94  
3            3    1056.18          356.70      1543.57     1077.35    466.22  
4            1    3902.00         6500.20     11364.40     8663.60   2700.80  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [154]:
# Check your work using the totals above

orders = [2742071, 2173913, 6128929]

order_totals = df[df["order_id"].isin(orders)].groupby("order_id")["total_price"].sum()

for order_id in orders:
    total_price = order_totals.get(order_id)
    print(f"Total price for order {order_id} is ${total_price:,.2f}")

Total price for order 2742071 is $152,811.89
Total price for order 2173913 is $162,388.71
Total price for order 6128929 is $923,441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [155]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_client_spend = (
    df[df['client_id'].isin(top_5)]
    .groupby('client_id')['total_price']
    .sum()
    .sort_values(ascending=False)
)

for client_id, total_spend in top_client_spend.items():
    print(f"Client {client_id}'s total spend is ${total_spend:,.2f}")


Client 24741's total spend is $82,268,892.04
Client 38378's total spend is $12,906,550.87
Client 66037's total spend is $10,259,514.80
Client 46820's total spend is $9,743,794.33
Client 33615's total spend is $8,377,308.53


In [156]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 

top_5_df = df[df['client_id'].isin(top_5)]
summary_df = top_5_df.groupby('client_id').agg(
    total_units_purchased = ('qty', 'sum'),
    total_shipping_price = ('shipping_price', 'sum'),
    total_revenue = ('total_price', 'sum'),
    total_profit = ('profit', 'sum')
).sort_values(by='total_revenue', ascending=False)

summary_df



total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
24741                     239862            5126448.37    82268892.04   
38378                      73667            3429455.40    12906550.87   
66037                      43018            1395151.85    10259514.80   
46820                      75768            1601448.84     9743794.33   
33615                      64313            1828984.89     8377308.53   

           total_profit  
client_id                
24741       36579992.33  
38378        3271829.89  
66037        3255031.82  
46820        2736602.69  
33615        2201994.62

In [157]:
print(summary_df.columns)

Index(['total_units_purchased', 'total_shipping_price', 'total_revenue',
       'total_profit'],
      dtype='object')


In [158]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
money_columns = ['total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions.
def currency_format(amount):
    return f"${amount/1000000:.2f}"


# Apply the currency_format_millions function to only the money columns. 
for column in money_columns:
    summary_df[column] = summary_df[column].apply(currency_format)


# Rename the columns to reflect the change in the money format. 
summary_df = summary_df.rename(columns={
    'total_shipping_price': 'Total Shipping Price (Millions)',
    'total_revenue': 'Total Revenue (Millions)',
    'total_profit': 'Total Profit (Millions)'
})

summary_df


total_units_purchased Total Shipping Price (Millions)  \
client_id                                                          
24741                     239862                           $5.13   
38378                      73667                           $3.43   
66037                      43018                           $1.40   
46820                      75768                           $1.60   
33615                      64313                           $1.83   

          Total Revenue (Millions) Total Profit (Millions)  
client_id                                                   
24741                       $82.27                  $36.58  
38378                       $12.91                   $3.27  
66037                       $10.26                   $3.26  
46820                        $9.74                   $2.74  
33615                        $8.38                   $2.20

In [161]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.

final_df = summary_df.sort_values(by='Total Profit (Millions)', ascending=False)
final_df

total_units_purchased Total Shipping Price (Millions)  \
client_id                                                          
24741                     239862                           $5.13   
38378                      73667                           $3.43   
66037                      43018                           $1.40   
46820                      75768                           $1.60   
33615                      64313                           $1.83   

          Total Revenue (Millions) Total Profit (Millions)  
client_id                                                   
24741                       $82.27                  $36.58  
38378                       $12.91                   $3.27  
66037                       $10.26                   $3.26  
46820                        $9.74                   $2.74  
33615                        $8.38                   $2.20

In [ ]:
'''The firm seems to have a whale of a client (id 24741), generating $82.4 million in revenue and $36.6 million in profit from that client alone.
The next biggest client only accounts for $12.9 million in revenue. 
'''